_posts/2021-04-05-maven-bom.md

```diff
@@ -0,0 +1,316 @@
+---
+title: "Resolving Maven Dependencies with a Bill of Materials (BOM)"
+categories: [java]
+date: 2021-04-05 00:00:00 +1100
+modified: 2021-04-05 00:00:00 +1100
+author: bakic
+excerpt: "Resolving dependencies in Maven using a Bill Of Materials (BOM)."
+image:
+  auto: 0045-checklist
+---
+
+As Java developers, we may maintain many applications using Maven for their dependency management. These applications need upgrades from time to time to be up to date and to add new features or security updates. 
+
+This easy task - updating dependencies' versions - can easily turn out to become a nightmare because of conflicts between certain dependencies. The resolution of these dependency conflicts can take a lot of time.
+
+**To make dependency management easier, we can use the Bill of Materials (BOM)**, a feature that offers easier and safer dependency management.
+
+In this article, we are going to look at dependency management in Maven and look at the BOM with some examples.
+
+## Direct vs. Transitive Dependencies
+
+Let's imagine we write some business code that requires logging the output, using some String utilities, or securing the application. This logic can be implemented in our project, or we can use a library instead. It often makes sense to use existing libraries to minimize the amount of code we need to write ourselves.
+
+The use of libraries encourages reuse since we will rely on other libraries that solve problems similar to ours: these libraries are our dependencies.
+
+There are two types of dependencies in Maven:
+
+- **direct dependencies**:  dependencies that are explicitly included in our Project Object Model (`pom.xml`) file in the `<dependencies>` section. They can be added using the `<dependency>` tag. Here is an example of a logging library added to a `pom.xml` file:
+
+  ```xml
+  <dependencies>
+      <dependency>
+          <groupId>log4j</groupId>
+          <artifactId>log4j</artifactId>
+          <version>1.2.17</version>
+      </dependency>
+  </dependencies>
+  ```
+
+- **transitive dependencies**: a project that we include as a dependency in our project, like the logging library above, can declare its own dependencies in a `pom.xml` file. **These dependencies are then considered *transitive* dependencies to our project**. When Maven pulls a direct dependency, it also pulls its transitive dependencies.
+
+## Transitive Dependencies with Maven
+
+Now that we have an overview of the different dependency types in Maven, let's see in detail how Maven deals with transitive dependencies in a project.
+
+As an example, we'll look at two dependencies from the Spring Framework: `spring-context` and `spring-security-web`.
+
+In the `pom.xml` file we add them as direct dependencies, purposefully selecting two different version numbers:
```

Comment:
We may choose another word instead of **purposefully**, it is not a common word for readers I think. Maybe **deliberately** or **intentionally**?

Reply:
yes, good idea! I will change that